lib/colonel_kurtz/block_type.ex

```diff
@@ -1,30 +1,65 @@
 defmodule ColonelKurtz.BlockType do
+  @moduledoc """
+  The BlockType module defines a macro that is used to mix in the Block Type
+  behavior. Block Types are embedded Ecto Schemas that conform to the
+  Validatable protocol.
+
+  A Block is of some type (e.g. "Image Block") and have a well-defined schema
+  for the attributes they require. That schema is specified as a dynamically
+  generated nested module within the using module and is defined by a user call
+  to the `defattributes` macro that specifies the schema for the Blocks' Content.
+  The Block Type schema embeds_one of the Content module.
+  """
+
+  import Ecto.Changeset, only: [add_error: 4]
+  import ColonelKurtz.BlockTypeContent, only: [defcontentmodule: 1]
+
+  alias ColonelKurtz.EctoHelpers
+
+  @type changeset :: Ecto.Changeset.t
+
+  @type block :: %{
+    required(:type) => binary,
+    required(:content) => map,
+    required(:blocks) => list(block)
+  }
+
+  @type block_struct :: %{
+    :__struct__ => atom,
+    required(:block_id) => nil | binary,
+    required(:type) => binary,
+    required(:content) => map,
+    required(:blocks) => list(block)
+  }
+
   defmacro __using__(opts) do
     quote do
       use Ecto.Schema
       @primary_key {:block_id, :binary_id, autogenerate: false}
 
-      import unquote(__MODULE__)
+      import unquote(__MODULE__), only: [
+        defattributes: 1,
+        attributes_from_params: 2,
+        lift_content_errors: 1
+      ]
+
       import Ecto.Changeset
       import ColonelKurtz.Validation, only: [validate_blocks: 3]
 
-      alias ColonelKurtz.ValidatableBlock
-      alias ColonelKurtz.EctoHelpers
       alias ColonelKurtz.BlockTypes
       alias ColonelKurtz.EctoBlocks
+      alias ColonelKurtz.Validatable
 
       @before_compile unquote(__MODULE__)
       @after_compile unquote(__MODULE__)
 
-      # A BlockType has a set of explicitly defined attributes (`block.content`)
-      @attributes []
-      @block_attributes [:block_id, :type, :content, :blocks]
       @block_type unquote(opts[:type])
-
+      @block_attributes [:block_id, :type, :content, :blocks]
+      @content_attributes []
       @content_module Module.concat(__MODULE__, Content)
 
       @derive [
-        ValidatableBlock,
+        Validatable,
         {Jason.Encoder, only: @block_attributes}
```

Comment:
I don't know how to replace our derived protocol for `Jason.Encoder` if we make the json library configurable. Something to come back to and figure out.